src/QirRuntime/test/QIR-static/qsharp/qir-test-other.qs

```diff
@@ -0,0 +1,27 @@
+// Copyright (c) Microsoft Corporation.
+// Licensed under the MIT License.
+
+namespace Microsoft.Quantum.Testing.QIR.Other {
+
+    open Microsoft.Quantum.Bitwise;
+
+    function ParityTest() : Int {
+        //function Parity (a : Int) : Int
+        if 0 != Parity(0)                       { return  1; }
+        if 1 != Parity(1)                       { return  2; }
+        if 1 != Parity(2)                       { return  3; }
+        if 0 != Parity(3)                       { return  4; }
+        if 0 != Parity(0xFF)                    { return  5; }
+        if 1 != Parity(0x100)                   { return  6; }
+        if 0 != Parity(0xFFFF)                  { return  7; }
+        if 1 != Parity(0x10000)                 { return  8; }
+        if 1 != Parity(0x7F00000000000000)      { return  9; }
+        if 0 != Parity(0x0F00000000000000)      { return 10; }
+                                                       
+        if 0 != Parity(-1)                      { return 11; }    // 0xFFFFFFFFFFFFFFFF
```

Comment:
The current behavior of `Parity` is to fail if its input is negative:

![image](https://user-images.githubusercontent.com/31516/109075705-6113b280-76ae-11eb-951c-d9a31dca0d84.png)

Your implementation generalizes that to return the parity of the twos-complement representation of negative inputs, which could be reasonable, but should at least be specified in the documentation comment for `Parity`.

Separately, it looks like [the language spec is currently a bit vague](https://github.com/microsoft/qsharp-language/blob/main/Specifications/Language/3_Expressions/BitwiseExpressions.md) on what `&&&` returns for negative integers. @bettinaheim, would it make sense to mention explicitly that the twos-complement representation is used in that case?

Reply:
Hmm... :-/ It is rather surprising to find out about any current behavior of `Parity()`. My understanding was that `Parity()` is currently [defined `intrinsic`](https://github.com/microsoft/qsharp-runtime/blob/d9d08086d5c655afb0ad9ccf9f7c0870589ea7ef/src/Simulation/QSharpFoundation/Bitwise/Bitwise.qs#L85) and has no implementation. I assume, you might mean something related to generation of the C# code (from Q#), which I know nearly nothing about. 
It is also suspicious to see the `UInt64` in the end of your screenshot. It makes me think that currently the signature has a parameter of the _unsigned_ 64-bit integer type (as a opposed to _signed_ 64-bit integer - `Int`), which contradicts the [documentation](https://docs.microsoft.com/en-us/qsharp/api/qsharp/microsoft.quantum.bitwise.parity).
Another curious observation I had when I was writing a test. I got a compiler error for the fragment
`Parity(0xFFFFFFFFFFFFFFFF)` (64-bit value consisting of 1s in all bit positions). The error was telling that the value does not fit in the data type, or something like that. This made me think that the compiler treaded the value as an _unsigned_ 64-bit integer, which really does not fit in the _signed_ 64-bit type (at least in two's complement representation).
To summarize, I have a suspicion, that the failure described by you, happens _before_ the actual call to `Parity()` (and may be happens during compilation).

I'm updating the documentation comment for `Parity()` to the following:
```q#
    /// Returns the bitwise PARITY of an integer 
    /// (1 if its [two's complement](https://en.wikipedia.org/wiki/Signed_number_representations#Two's_complement) 
    /// representation contains odd number of ones and 0 otherwise).
```

Reply:
The implementation is here:
https://github.com/microsoft/qsharp-runtime/blob/d9d08086d5c655afb0ad9ccf9f7c0870589ea7ef/src/Simulation/QSharpFoundation/Bitwise/Bitwise.cs#L49
And indeed, the first thing it does is convert to a UInt64.

Reply:
@kuzminrobin: The current implementation is provided by the C# runtime, as you say. In particular, the implementation at [line 49 of `Bitwise.cs`](https://github.com/microsoft/qsharp-runtime/blob/cgranade/experimental/opensim/src/Simulation/QsharpFoundation/Bitwise/Bitwise.cs#L49) starts by converting the input to `Parity` (Q# `Int`, represented in the C# runtime as `long`) to a C# `ulong` with a call to `System.Convert.ToUInt64`. Thus, the error message in the screenshot is not from before `Parity`, but is an exception in the C# runtime that is converted to a Q# `fail`.

Reply:
@swernli: jinx...!

Reply:
@swernli, @cgranade,
Thank you for the link and explanation, very informative!
As I understand, the reasons behind the cast/conversion to the _unsigned_ integer type were the same as [here](https://github.com/kuzminrobin/code_review_notes/blob/master/cpp_design_bookmarks.md#know-the-danger-of-overflowing-and-underflowing-the-signed-types) (still applicable to the implementations prior to C++20).

Grand Total:
I'm adding to the comment that the two's complement representation is used.
Changing the implementation to the one that was in `.cs`.